File: Matlab_Exercises/Exercise_3/D_Projection_and_Spatial_Structures.m


%% D. Compute Projection.

% Given the temporal Basis for the POD and the mPOD we compute their
% spatial structures

clear all; clc; close all

% Load Data
load('Data.mat')
[n_x,n_y]=size(Xg);
n_s=size(D,1);
% Load mPOD basis
load('Psis_mPOD.mat','PSI_M')

% Load POD basis
load('Psis_POD.mat','PSI_P','Sigma_P')

 
%% Compute the spatial basis for mPOD and POD
% Projection from Eq. 2.6
R=size(PSI_M,2);
PHI_M_SIGMA_M=D*PSI_M;
% Initialize the output
PHI_M=zeros([n_s,R]);
SIGMA_M=zeros([R,R]);

 for i=1:1:R
disp(['Completing mPOD Mode ',num2str(i)]);
% Normalize the columns of C to get spatial modes
PHI_M(:,i) = PHI_M_SIGMA_M(:,i)/norm(PHI_M_SIGMA_M(:,i));
% Assign the norm as amplitude
SIGMA_M(i,i) = norm(PHI_M_SIGMA_M(:,i));
end

% Sort the amplitudes in decreasing order
[Sort_SM,Perm]=sort(diag(SIGMA_M),'descend');

Phi_M = PHI_M(:,Perm); % Spatial Basis mPOD
Psi_M = PSI_M(:,Perm); % Temporal Basis mPOD
Sigma_VV = diag(SIGMA_M);
Sigma_VV = Sigma_VV(Perm);
Sigma_M = diag(Sigma_VV); % Amplitude for mPOD Basis



% First 3 Modes Results mPOD
HFIG=figure(1);
HFIG.Units='normalized';
HFIG.Position=[0.1 0.1 0.65 0.75];
HFIG.Name='Results mPOD Analysis- Part 1';
for j=1:3
subplot(3,2,2*j-1)
% Reconstruct the Omega from the columns
Omega=reshape(Phi_M(:,j+1),[n_x,n_y]);

pcolor(Xg,Yg,Omega);
shading interp
daspect([1 1 1])
ylim([-19 19])  
  xlim([-19 19])
 set(gca, ...
      'Fontname', 'Palatino Linotype', ...
      'Fontsize', 16, ...
       'Box'         , 'off'     , ...   
        'LineWidth'   , 1         )
 % Label Information
 xlabel('$x[-]$','Interpreter','Latex')
 ylabel('$y[-]$','Interpreter','Latex')
 title(['Spatial Structure ',num2str(j+1)],'Interpreter','Latex')
 set(gcf,'color','white')
end



Fs=1/dt; % Sampling frequency
Freq = [-n_t/2:1:n_t/2-1]*(Fs)*1/n_t; % Frequency Axis


for j=1:3
subplot(3,2,2*j)
plot(Freq,abs(fftshift(fft(Psi_M(:,j+1)))),'linewidth',1.5)
 set(gca, ...
      'Fontname', 'Palatino Linotype', ...
      'Fontsize', 16, ...
       'Box'         , 'off'     , ...   
        'LineWidth'   , 1         )
xlim([0 20])
 % Label Information
 xlabel('$f[-]$','Interpreter','Latex')
 ylabel('$\widehat{\psi}_{\mathcal{M}}$','Interpreter','Latex')
 title(['Frequency in the T. Structure ',num2str(j+1)],'Interpreter','Latex')
 set(gcf,'color','white')
 end

print(HFIG,'Results_mPOD_1.png','-dpng')



% First 3 Modes Results mPOD
HFIG=figure(2);
HFIG.Units='normalized';
HFIG.Position=[0.1 0.1 0.65 0.75];
HFIG.Name='Results mPOD Analysis- Part 2';
for j=1:3
subplot(3,2,2*j-1)
% Reconstruct the Omega from the columns
Omega=reshape(Phi_M(:,j+4),[n_x,n_y]);

pcolor(Xg,Yg,Omega);
shading interp
daspect([1 1 1])
ylim([-19 19])  
  xlim([-19 19])
 set(gca, ...
      'Fontname', 'Palatino Linotype', ...
      'Fontsize', 16, ...
       'Box'         , 'off'     , ...   
        'LineWidth'   , 1         )
 % Label Information
 xlabel('$x[-]$','Interpreter','Latex')
 ylabel('$y[-]$','Interpreter','Latex')
 title(['Spatial Structure ',num2str(j+1)],'Interpreter','Latex')
 set(gcf,'color','white')
end



Fs=1/dt; % Sampling frequency
Freq = [-n_t/2:1:n_t/2-1]*(Fs)*1/n_t; % Frequency Axis


for j=1:3
subplot(3,2,2*j)
plot(Freq,abs(fftshift(fft(Psi_M(:,j+4)))),'linewidth',1.5)
 set(gca, ...
      'Fontname', 'Palatino Linotype', ...
      'Fontsize', 16, ...
       'Box'         , 'off'     , ...   
        'LineWidth'   , 1         )
xlim([0 20])
 % Label Information
 xlabel('$f[-]$','Interpreter','Latex')
 ylabel('$\widehat{\psi}_{\mathcal{M}}$','Interpreter','Latex')
 title(['Frequency in the T. Structure ',num2str(j+1)],'Interpreter','Latex')
 set(gcf,'color','white')
 end

print(HFIG,'Results_mPOD_2.png','-dpng')


% Compute the spatial basis for the POD
load('Psis_POD.mat','PSI_P','Sigma_P')
R=size(PSI_P,2);
PHI_P_SIGMA_P=D*PSI_P;
% Initialize the output
Phi_P=zeros([n_s,R]);

 for i=1:1:R
disp(['Completing POD Mode ',num2str(i)]);
% Normalize the columns of C to get spatial modes
Phi_P(:,i) = PHI_P_SIGMA_P(:,i)/Sigma_P(i,i);
% The POD does not need a normalization for the amplitude!
end


% First 3 Modes Results POD
HFIG=figure(3);
HFIG.Units='normalized';
HFIG.Position=[0.1 0.1 0.65 0.75];
HFIG.Name='Results POD Analysis- Part 1';
for j=1:3
subplot(3,2,2*j-1)
% Reconstruct the fields from the columns

Omega=reshape(Phi_M(:,j+4),[n_x,n_y]);

pcolor(Xg,Yg,Omega);
shading interp
daspect([1 1 1])
ylim([-19 19])  
  xlim([-19 19])
 set(gca, ...
      'Fontname', 'Palatino Linotype', ...
      'Fontsize', 16, ...
       'Box'         , 'off'     , ...   
        'LineWidth'   , 1         )
 % Label Information
 xlabel('$x[-]$','Interpreter','Latex')
 ylabel('$y[-]$','Interpreter','Latex')
 title(['Spatial Structure ',num2str(j+1)],'Interpreter','Latex')
 set(gcf,'color','white')
end



Fs=1/dt; % Sampling frequency
Freq = [-n_t/2:1:n_t/2-1]*(Fs)*1/n_t; % Frequency Axis


for j=1:3
subplot(3,2,2*j)
plot(Freq,abs(fftshift(fft(PSI_P(:,j+1)))),'linewidth',1.5)
 set(gca, ...
      'Fontname', 'Palatino Linotype', ...
      'Fontsize', 16, ...
       'Box'         , 'off'     , ...   
        'LineWidth'   , 1         )
xlim([0 20])
 % Label Information
 xlabel('$f[-]$','Interpreter','Latex')
 ylabel('$\widehat{\psi}_{\mathcal{P}}$','Interpreter','Latex')
 title(['Frequency in the T. Structure ',num2str(j+1)],'Interpreter','Latex')
 set(gcf,'color','white')
 end

print(HFIG,'Results_POD_1.png','-dpng')

% First 3 Modes Results POD
HFIG=figure(4);
HFIG.Units='normalized';
HFIG.Position=[0.1 0.1 0.65 0.75];
HFIG.Name='Results POD Analysis- Part 2';
for j=1:3
subplot(3,2,2*j-1)
% Reconstruct the fields from the columns
Omega=reshape(Phi_P(:,j+4),[n_x,n_y]);

pcolor(Xg,Yg,Omega);
shading interp
daspect([1 1 1])
ylim([-19 19])  
  xlim([-19 19])
 set(gca, ...
      'Fontname', 'Palatino Linotype', ...
      'Fontsize', 16, ...
       'Box'         , 'off'     , ...   
        'LineWidth'   , 1         )
 % Label Information
 xlabel('$x[-]$','Interpreter','Latex')
 ylabel('$y[-]$','Interpreter','Latex')
 title(['Spatial Structure ',num2str(j+4)],'Interpreter','Latex')
 set(gcf,'color','white')
end


Fs=1/dt; % Sampling frequency
Freq = [-n_t/2:1:n_t/2-1]*(Fs)*1/n_t; % Frequency Axis


for j=1:3
subplot(3,2,2*j)
plot(Freq,abs(fftshift(fft(PSI_P(:,j+4)))),'linewidth',1.5)
 set(gca, ...
      'Fontname', 'Palatino Linotype', ...
      'Fontsize', 16, ...
       'Box'         , 'off'     , ...   
        'LineWidth'   , 1         )
xlim([0 20])
 % Label Information
 xlabel('$f[-]$','Interpreter','Latex')
 ylabel('$\widehat{\psi}_{\mathcal{P}}$','Interpreter','Latex')
 title(['Frequency in the T. Structure ',num2str(j+4)],'Interpreter','Latex')
 set(gcf,'color','white')
 end

print(HFIG,'Results_POD_2.png','-dpng')

File: Python_Exercises/Exercise_4/D_Projection_and_Spatial_Structures.py
# -*- coding: utf-8 -*-
"""
Created on Wed Jan  1 14:01:17 2020

@author: mendez
"""


## D. Compute Projection.

# Given the temporal Basis for the POD and the mPOD we compute their
# spatial structures


import numpy as np
import matplotlib.pyplot as plt
from Others import Plot_Field
# Load Data
data = np.load('Data.npz')
D=data['D']; t=data['t']; dt=data['dt']; n_t=data['n_t'];
Fs=1/dt
Xg=data['Xg']; Yg=data['Yg']; X_S=data['X_S']; Y_S=data['Y_S']
n_y,n_x=Yg.shape; nxny=n_x*n_y; n_s=nxny*2

# Load mPOD basis
data=np.load('Psis_mPOD.npz')
Psi_M=data['PSI_M']
 
# Compute the spatial basis for mPOD 
# Projection from Eq. 2.6
R=Psi_M.shape[1];
PHI_M_SIGMA_M=np.dot(D,(Psi_M))
# Initialize the output
PHI_M=np.zeros((n_s,R))
SIGMA_M=np.zeros((R))

for i in range(0,R):     
  print('Completing mPOD Mode '+str(i))
  #Assign the norm as amplitude
  SIGMA_M[i] = np.linalg.norm(PHI_M_SIGMA_M[:,i])
  #Normalize the columns of C to get spatial modes
  PHI_M[:,i] = PHI_M_SIGMA_M[:,i]/SIGMA_M[i]
  
Indices=np.flipud(np.argsort(SIGMA_M)) # find indices for sorting in decreasing order
Sorted_Sigmas=SIGMA_M[Indices] # Sort all the sigmas
Phi_M=PHI_M[:,Indices] # Sorted Spatial Structures Matrix
Psi_M=Psi_M[:,Indices] # Sorted Temporal Structures Matrix
Sigma_M=Sorted_Sigmas # Sorted Amplitude Matrix
  
# Show some exemplary modes for mPOD. 

for j in range(1,6):
 plt.close(fig='all') 
 fig, ax3= plt.subplots(figsize=(5,7))   
 ax=plt.subplot(2,1,1)
 V_X=Phi_M[0:nxny,j-1]
 V_Y=Phi_M[nxny::,j-1]
 Plot_Field(X_S,Y_S,V_X,V_Y,True,2,0.8)
 ax.set_aspect('equal') # Set equal aspect ratio
 ax.set_xticks(np.arange(0,40,10))
 ax.set_yticks(np.arange(10,30,5))
 ax.set_xlim([0,35])
 ax.set_ylim([10,29])
 ax.invert_yaxis() # Invert Axis for plotting purpose
 String_y='$\phi_{\mathcal{P}'+str(j)+'}$'
 plt.title(String_y,fontsize=18)
# plt.tight_layout(pad=1, w_pad=0.5, h_pad=1.0)
 
 ax=plt.subplot(2,1,2)
 Signal=Psi_M[:,j-1]
 s_h=np.abs((np.fft.fft(Signal-Signal.mean())))
 Freqs=np.fft.fftfreq(int(n_t))*Fs
 plt.plot(Freqs*(4/1000)/6.5,s_h,'-',linewidth=1.5)
 plt.xlim(0,0.38)    
 plt.xlabel('$St[-]$',fontsize=18)
 String_y='$\widehat{\psi}_{\mathcal{M}'+str(j)+'}$'
 plt.ylabel(String_y,fontsize=18)
 plt.tight_layout(pad=1, w_pad=0.5, h_pad=1.0)
 Name='mPOD_Mode_'+str(j)+'.png'
 print(Name+' Saved')
 plt.savefig(Name, dpi=300)  



# Compute the spatial basis for the POD

# Load POD basis
data=np.load('Psis_POD.npz')
Psi_P=data['Psi_P']
Sigma_P=data['Sigma_P']

# Projection from Eq. 2.6
R=Psi_P.shape[1];
PHI_P_SIGMA_P=np.dot(D,(Psi_P))
# Initialize the output
Phi_P=np.zeros((n_s,R))

for i in range(0,R):     
  print('Completing POD Mode '+str(i))
  #Normalize the columns of C to get spatial modes
  Phi_P[:,i] = PHI_P_SIGMA_P[:,i]/Sigma_P[i]
  
for j in range(1,10):
 plt.close(fig='all') 
 fig, ax3= plt.subplots(figsize=(5,7))   
 ax=plt.subplot(2,1,1)
 V_X=Phi_P[0:nxny,j-1]
 V_Y=Phi_P[nxny::,j-1]
 Plot_Field(X_S,Y_S,V_X,V_Y,True,2,0.8)
 ax.set_aspect('equal') # Set equal aspect ratio
 ax.set_xticks(np.arange(0,40,10))
 ax.set_yticks(np.arange(10,30,5))
 ax.set_xlim([0,35])
 ax.set_ylim([10,29])
 ax.invert_yaxis() # Invert Axis for plotting purpose
 String_y='$\phi_{\mathcal{P}'+str(j)+'}$'
 plt.title(String_y,fontsize=18)
# plt.tight_layout(pad=1, w_pad=0.5, h_pad=1.0)
 
 ax=plt.subplot(2,1,2)
 Signal=Psi_P[:,j-1]
 s_h=np.abs((np.fft.fft(Signal-Signal.mean())))
 Freqs=np.fft.fftfreq(int(n_t))*Fs
 plt.plot(Freqs*(4/1000)/6.5,s_h,'-',linewidth=1.5)
 plt.xlim(0,0.38)    
 plt.xlabel('$St[-]$',fontsize=18)
 String_y='$\widehat{\psi}_{\mathcal{P}'+str(j)+'}$'
 plt.ylabel(String_y,fontsize=18)
 plt.tight_layout(pad=1, w_pad=0.5, h_pad=1.0)
 Name='POD_Mode_'+str(j)+'.png'
 print(Name+' Saved')
 plt.savefig(Name, dpi=300)  
 

plt.close(fig='all')
# Amplitude Modes POD vs mPOD
RR=np.arange(1,Sigma_P.shape[0]+1)
## DFT Spectra and convergence
fig, ax = plt.subplots(figsize=(6,3))
plt.rc('text', usetex=True)     
plt.rc('font', family='serif')
plt.rc('xtick',labelsize=12)
plt.rc('ytick',labelsize=12)
sigma_P_v=(Sigma_P)/(n_s*n_t)*1000
sigma_M_v=(Sigma_M)/(n_s*n_t)*1000
plt.plot(RR,sigma_P_v,'ko',label='POD')
plt.plot(RR,sigma_M_v,'rs',label='mPOD')
plt.legend(fontsize=20,loc='upper right') 
plt.xlabel('$r$',fontsize=12)
plt.xlim(0.5,1001)
plt.xscale('log')
plt.ylabel('$\sigma_{\mathcal{P}r}/\sigma_{\mathcal{M}r}*1000$',fontsize=14)
plt.title('POD vs mPOD Amplitudes ',fontsize=16)
plt.tight_layout(pad=0.6, w_pad=0.3, h_pad=0.8)
#pos1 = ax.get_position() # get the original position 
#pos2 = [pos1.x0 + 0.01, pos1.y0 + 0.01,  pos1.width *0.95, pos1.height *0.95] 
#ax.set_position(pos2) # set a new position
plt.savefig('POD_mPOD_Amplitude.png', dpi=200)     
plt.close(fig)






